sdks/go/pkg/beam/core/runtime/harness/session.go

```diff
@@ -213,3 +186,59 @@ func recordFooter() error {
 		},
 	})
 }
+
+// CaptureHook writes the messaging content consumed and
+// produced by the worker, allowing the data to be used as
+// an input for the session runner. Since workers can exist
+// in a variety of environments, this allows the runner
+// to tailor the behavior best for its particular needs.
+type CaptureHook io.WriteCloser
+
+// CaptureHookFactory produces a CaptureHook from the supplied
+// options.
+type CaptureHookFactory func([]string) CaptureHook
+
+var captureHookRegistry = make(map[string]CaptureHookFactory)
+var enabledCaptureHook string
+
+func init() {
+	hf := func(opts []string) hooks.Hook {
+		return hooks.Hook{
+			Init: func(_ context.Context) error {
```

Comment:
<!--new_thread; commit:abd6cffe5451804ffdc2a1902dbdf7f3b588c57f; resolved:0-->
As implemented, all capture hooks run at worker initialization time.

Is there any further implementation needed?

Reply:
The hooks are invoked inside the harness package via the entry points 'recordInstructionRequest' and 'recordInstructionResponse' that apply a serialization format and then invoke the registered hooks.